### PR TITLE
Support service calls with return values in codegen

### PR DESCRIPTION
**Pull request by helto4real**
## Proposed change

This PR adds support for calling services with return values in the code generator.

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue:
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code compiles without warnings (code quality check)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/net-daemon/docs

### CODE REVIEW

File: src/HassModel/NetDaemon.HassModel.Tests/CodeGenerator/ServicesGeneratorTest.cs [hunk 1]
```diff
@@ -247,6 +247,56 @@ public void Run(Entities entities, Services services)
...
+                            Optional = true,
+                        }
+                    },
+                ],
+            }
+        };
+
+        // Act:
+        var code = CodeGenTestHelper.GenerateCompilationUnit(_settings, readOnlyCollection, hassServiceDomains);
+        code.ToString().Should().NotContain("DumbthingsEntityExtensionMethods",
```
Comment (FrankBakkerNl) [inline src/HassModel/NetDaemon.HassModel.Tests/CodeGenerator/ServicesGeneratorTest.cs:341]: Looks like copied from another test, but this assert is not relevant here

File: src/HassModel/NetDaemon.HassModel.Tests/CodeGenerator/ServicesGeneratorTest.cs [hunk 2]
```diff
@@ -247,6 +247,56 @@ public void Run(Entities entities, Services services)
     }
 
     [Fact]
+    public void TestServiceWithReturnvalueWithoutAnyMethods()
+    {
+        var readOnlyCollection = new HassState[] {
+            new() { EntityId = "weather.hometown" },
+        };
+
+        var hassServiceDomains = new HassServiceDomain[] {
```
Comment (FrankBakkerNl) [inline src/HassModel/NetDaemon.HassModel.Tests/CodeGenerator/ServicesGeneratorTest.cs:320]: I think for some other tests I switched to having the json format in the test and parse that, iso the object initializers. I think those are a bit more clear

File: src/HassModel/NetDaemon.HassModel.CodeGenerator/CodeGeneration/ServicesGenerator.cs [hunk 3]
```diff
@@ -132,6 +145,31 @@ private static IEnumerable<MemberDeclarationSyntax> GenerateServiceMethod(string
                 .WithSummaryComment(service.Description)
                 .AppendTrivia(targetComment)
                 .AppendParameterComments(serviceArguments);
+            if (service.Response is not null)
```
Comment (FrankBakkerNl) [inline src/HassModel/NetDaemon.HassModel.CodeGenerator/CodeGeneration/ServicesGenerator.cs:148]: Seems like this method is getting pretty long, maybe split out sostuff like for the other one

File: src/HassModel/NetDaemon.HassModel.CodeGenerator/CodeGeneration/ExtensionMethodsGenerator.cs [hunk 4]
```diff
@@ -69,16 +69,33 @@ private static IEnumerable<MemberDeclarationSyntax> GenerateExtensionMethodsForS
 
         if (serviceArguments is null)
         {
-            yield return ExtensionMethodWithoutArguments(service, serviceName, entityTypeName);
-            yield return ExtensionMethodWithoutArguments(service, serviceName, enumerableTargetTypeName);
+            if (service.Response is not null)
```
Comment (FrankBakkerNl) [inline src/HassModel/NetDaemon.HassModel.CodeGenerator/CodeGeneration/ExtensionMethodsGenerator.cs:72]: This if else and all the slightly different methods is getting a bit out of hand. Maybe we could make a more generic method to generate a method based on some bool parameters.

File: src/HassModel/NetDaemon.HassModel.Tests/CodeGenerator/ServicesGeneratorTest.cs (hunk 1 above)
Comment (helto4real) [reply src/HassModel/NetDaemon.HassModel.Tests/CodeGenerator/ServicesGeneratorTest.cs:341]: That is very true my friend :)

File: src/HassModel/NetDaemon.HassModel.CodeGenerator/CodeGeneration/ExtensionMethodsGenerator.cs (hunk 4 above)
Comment (helto4real) [reply src/HassModel/NetDaemon.HassModel.CodeGenerator/CodeGeneration/ExtensionMethodsGenerator.cs:72]: Yea I was thinking about that too, just wanted to get something working first...

File: src/HassModel/NetDaemon.HassModel.Tests/CodeGenerator/ServicesGeneratorTest.cs (hunk 2 above)
Comment (helto4real) [reply src/HassModel/NetDaemon.HassModel.Tests/CodeGenerator/ServicesGeneratorTest.cs:320]: Yea I realized that after I wrote the tests but not sure I want to refactor these now.

File: src/HassModel/NetDeamon.HassModel/Entities/EnumerableEntityExtensions.cs [hunk 5]
```diff
@@ -78,28 +78,64 @@
...
     }
-}
+    /// <summary>
+    /// Calls a service with a set of Entities as the target
+    /// </summary>
+    /// <param name="entities">IEnumerable of Entities for which to call the service</param>
+    /// <param name="service">Name of the service to call. If the Domain of the service is the same as the domain of the Entities it can be omitted</param>
+    /// <param name="data">Data to provide</param>
+    public static async Task<IEnumerable<JsonElement?>> CallServiceWithResponseAsync(this IEnumerable<IEntityCore> entities, string service, object? data = null)
+    {
```
Comment (FrankBakkerNl) [inline src/HassModel/NetDeamon.HassModel/Entities/EnumerableEntityExtensions.cs]: Did you find out if this also returns an array if you call it with one or zero entities?

**Review by FrankBakkerNl — APPROVED** [review]

**Pull request merged**